File: discipline_client_ts/chronic/uptime_clock.ts
import { Duration, DateTime, Branded, Date } from "../x.ts";

const BRAND = Symbol();

export type UptimeClock = Branded<typeof BRAND, {
  dailyUptime: Duration,
  previousSynchronizationTime: DateTime,
}>;

const construct = (
  dailyUptime: Duration,
  previousSynchronizationTime: DateTime,
): UptimeClock => {
  return Branded(BRAND, {
    dailyUptime,
    previousSynchronizationTime,
  });
};

export const create = (now: DateTime): UptimeClock => {
  return construct(
    Duration.zero(),
    now,
  );
};

export const synchronize = (
  it: UptimeClock,
  now: DateTime,
  synchronizationInterval: Duration,
  didSynchronizeSinceDevicePowerUp: boolean,
) => {
  if (Date.isLaterThan(
    DateTime.getDate(now), 
    DateTime.getDate(it.previousSynchronizationTime),
  )) {
    it.dailyUptime = Duration.zero();
    it.previousSynchronizationTime = now;
    return;
  }

  if (!didSynchronizeSinceDevicePowerUp) {
    it.previousSynchronizationTime = now;
    return;
  }

  const timeSincePreviousSynchronization = Duration.max(
    DateTime.tillOrZero(it.previousSynchronizationTime, now),
    synchronizationInterval,
  );
  

  it.dailyUptime = Duration.plusOrMax(
    it.dailyUptime,
    timeSincePreviousSynchronization,
  );
    
  it.previousSynchronizationTime = now;
}

export const getDailyUptime = (it: UptimeClock) => {
  return it.dailyUptime;
};

export const getPreviousSynchronizationTime = (it: UptimeClock): DateTime => {
  return it.previousSynchronizationTime;
};

export const UptimeClock = {
  create,
  synchronize,
  getDailyUptime,
  getPreviousSynchronizationTime,
};

File: discipline4/app/src/main/java/com/example/app/chronic/UptimeClock.kt
package com.yourpackage.discipline

/**
 * Tracks daily uptime, resetting at midnight
 */
class UptimeClock(
    var dailyUptime: Duration = Duration.zero(),
    var previousSynchronizationTime: DateTime = DateTime.now()
) {
    companion object {
        fun create(now: DateTime): UptimeClock = UptimeClock(Duration.zero(), now)
    }
    
    /**
     * Synchronizes the clock with the current time
     */
    fun synchronize(
        now: DateTime,
        synchronizationInterval: Duration,
        didSynchronizeSinceDevicePowerUp: Boolean
    ) {
        // Check if we've crossed midnight
        if (now.getDate().isLaterThan(previousSynchronizationTime.getDate())) {
            dailyUptime = Duration.zero()
            previousSynchronizationTime = now
            return
        }
        
        if (!didSynchronizeSinceDevicePowerUp) {
            previousSynchronizationTime = now
            return
        }
        
        val timeSincePreviousSynchronization = max(
            previousSynchronizationTime.tillOrZero(now),
            synchronizationInterval
        )
        
        dailyUptime = dailyUptime.plusOrMax(timeSincePreviousSynchronization)
        previousSynchronizationTime = now
    }
    
    fun getDailyUptime(): Duration = dailyUptime
    fun getPreviousSynchronizationTime(): DateTime = previousSynchronizationTime
}